src/language-js/printer-estree.js

```diff
@@ -3870,42 +3880,44 @@ function printFunctionParams(path, print, options, expandArg, printTypeParams) {
   const paramsField = fun.parameters ? "parameters" : "params";
   const isParametersInTestCall = isTestCall(parent);
   const shouldHugParameters = shouldHugArguments(fun);
-  const shouldExpandParameters =
-    expandArg &&
-    !(fun[paramsField] && fun[paramsField].some((n) => n.comments));
 
   const typeParams = printTypeParams
     ? printFunctionTypeParameters(path, options, print)
     : "";
 
-  let printed = [];
-  if (fun[paramsField]) {
-    const lastArgIndex = fun[paramsField].length - 1;
+  const parameters = [];
+  if (fun.this) {
+    parameters.push({ node: fun.this, doc: path.call(print, "this") });
+  }
 
-    printed = path.map((childPath, index) => {
-      const parts = [];
-      const param = childPath.getValue();
+  if (fun[paramsField]) {
+    path.each((childPath) => {
+      parameters.push({ node: childPath.getValue(), doc: print(childPath) });
+    }, paramsField);
+  }
 
-      parts.push(print(childPath));
+  const shouldExpandParameters =
+    expandArg && !parameters.some(({ node }) => node.comments);
 
-      if (index === lastArgIndex) {
-        if (fun.rest) {
-          parts.push(",", line);
-        }
-      } else if (
-        isParametersInTestCall ||
-        shouldHugParameters ||
-        shouldExpandParameters
-      ) {
-        parts.push(", ");
-      } else if (isNextLineEmpty(options.originalText, param, options.locEnd)) {
-        parts.push(",", hardline, hardline);
-      } else {
-        parts.push(",", line);
+  const printed = [];
+  const lastArgIndex = parameters.length - 1;
+  for (const [index, { node, doc }] of parameters.entries()) {
+    printed.push(doc);
+    if (index === lastArgIndex) {
+      if (fun.rest) {
+        printed.push(",", line);
       }
-
-      return concat(parts);
-    }, paramsField);
+    } else if (
+      isParametersInTestCall ||
+      shouldHugParameters ||
+      shouldExpandParameters
+    ) {
+      printed.push(", ");
+    } else if (isNextLineEmpty(options.originalText, node, options.locEnd)) {
+      printed.push(",", hardline, hardline);
+    } else {
+      printed.push(",", line);
+    }
   }
 
   if (fun.rest) {
```

Comment:
@fisker Overall, looks good. Did you consider printing `fun.rest` the same way by adding it to the `parameters` array first? And how about a `parameters` → `printedParameters` rename?

Reply:
I did think about it, but maybe later, it's not related to this PR.